File: scilab/modules/graphics/src/c/StringMatrix.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2006 - INRIA - Jean-Baptiste Silvy
 *
 * Copyright (C) 2012 - 2016 - Scilab Enterprises
 *
 * This file is hereby licensed under the terms of the GNU GPL v2.0,
 * pursuant to article 5.3.4 of the CeCILL v.2.1.
 * This file was originally licensed under the terms of the CeCILL v2.1,
 * and continues to be available under such terms.
 * For more information, see the COPYING file which you should have received
 * along with this program.
 *
 */

/*----------------------------------------------------------------------------------*/
/* COPYRIGHT INRIA 2006                                                                      */
/* Desc.   : Allocation and deletion and modifications of matrices of strings.               */
/*           The matrix is stored by colmuns like in Scilab                                  */
/*----------------------------------------------------------------------------------*/

#include <string.h>
#include <stdio.h>
#include "StringMatrix.h"
#include "sci_malloc.h"
#include "os_string.h"


/*----------------------------------------------------------------------------------*/
StringMatrix * newFullStringMatrix(char ** textMat, int nbRow, int nbCol)
{
    int i = 0;
    /* create the matrix */
    StringMatrix * newMat = newMatrix(nbRow, nbCol);

    /* copy each element */
    for (i = 0 ; i < nbRow * nbCol ; i++)
    {
        /* +1 for the /0 last character */
        newMat->data[i] = MALLOC((strlen(textMat[i]) + 1) * sizeof(char));
        strcpy(newMat->data[i], textMat[i]);
    }

    return newMat;
}
/*----------------------------------------------------------------------------------*/
StringMatrix * newEmptyStringMatrix(int nbRow, int nbCol)
{
    int i = 0;
    /* create the matrix */
    StringMatrix * newMat = newMatrix(nbRow, nbCol);

    /* copy each element */
    for (i = 0 ; i < nbRow * nbCol ; i++)
    {
        /* +1 for the /0 last character */
        newMat->data[i] = MALLOC((strlen("") + 1) * sizeof(char));
        strcpy(newMat->data[i], "");
    }

    return newMat;
}
/*----------------------------------------------------------------------------------*/
StringMatrix * copyStringMatrix(const StringMatrix * copyMat)
{
    return newFullStringMatrix((char **) copyMat->data, copyMat->nbRow, copyMat->nbCol);
}
/*----------------------------------------------------------------------------------*/
char * getStrMatElement(const StringMatrix * mat, int row, int col)
{
    return (char *) getMatElement(mat, row, col);
}
/*----------------------------------------------------------------------------------*/
char ** getStrMatData(const StringMatrix * mat)
{
    return (char **) getMatData(mat);
}
/*----------------------------------------------------------------------------------*/
void copyStrMatElement(StringMatrix * mat, int row, int col, const char * copyStr)
{
    char * changedString = (char *) mat->data[row + col * mat->nbRow] ; /* for speed */
    if (changedString != NULL)
    {
        FREE(changedString);
    }
    changedString = os_strdup(copyStr);
    mat->data[row + col * mat->nbRow] = changedString;
}
/*----------------------------------------------------------------------------------*/
void printStrMat(StringMatrix * mat)
{
    int i = 0;
    int nbRow = getMatNbRow(mat);
    int nbCol = getMatNbCol(mat);
    for (i = 0; i < nbRow; i++)
    {
        int j = 0;
        for (j = 0; j < nbCol; j++)
        {
            printf("%s ", getStrMatElement(mat, i, j));
        }
        printf("\n");
    }
}
/*----------------------------------------------------------------------------------*/

File: scilab/modules/renderer/src/java/org/scilab/modules/renderer/textDrawing/StringMatrix.java
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2007 - INRIA - Jean-Baptiste Silvy
 * desc : Matrix of string 
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */

package org.scilab.modules.renderer.textDrawing;

/**
 * Matrix of string
 * @author Jean-Baptiste Silvy
 */
public class StringMatrix {

	private int nbRow;
	private int nbCol;
	private String[][] data;
	
	
	/**
	 * Default constructor
	 */
	public StringMatrix() {
		nbRow = 0;
		nbCol = 0;
		data = null;
	}
	
	/**
	 * Construct a matrix of size nbRow x nbCol.
	 * @param nbRow number of row the matrix
	 * @param nbCol number of column of the matrix
	 */
	public StringMatrix(int nbRow, int nbCol) {
		data = new String[nbRow][nbCol];
		this.nbCol = nbCol;
		this.nbRow = nbRow;
	}
	
	/**
	 * Construct a new matrix from already given data.
	 * @param text array of strings conataining the data.
	 * @param nbRow number of row of the matrix.
	 * @param nbCol number of column of the matrix.
	 */
	public StringMatrix(String[] text, int nbRow, int nbCol) {
		setData(text, nbRow, nbCol);
	}
	
	/**
	 * Set the matrix data from an array of string
	 * @param text array of strings conataining the data.
	 * @param nbRow number of row of the matrix.
	 * @param nbCol number of column of the matrix.
	 */
	public void setData(String[] text, int nbRow, int nbCol) {
		if (data == null || this.nbRow != nbRow || this.nbCol != nbCol) {
			data = new String[nbRow][nbCol];
			this.nbCol = nbCol;
			this.nbRow = nbRow;
		}
		
		// scilab matrices are stored column wise
		for (int i = 0; i < nbRow; i++) {
			for (int j = 0; j < nbCol; j++) {
				data[i][j] = text[j * nbRow + i];
			}
		}
	}
	
	/**
	 * Change an element of the matrix.
	 * @param numRow row index
	 * @param numCol column index
	 * @param value new value to set
	 */
	public void setMatrixElement(int numRow, int numCol, String value) {
		data[numRow][numCol] = value;
	}
	
	/**
	 * @param numRow row index
	 * @param numCol column index
	 * @return element (numRow, numCol).
	 */
	public String getMatrixElement(int numRow, int numCol) {
		return data[numRow][numCol];
	}
	
	/**
	 * @return number of column of the matrix
	 */
	public int getNbCol() {
		return nbCol;
	}
	
	/**
	 * @return number of row of the matrix.
	 */
	public int getNbRow() {
		return nbRow;
	}
	
	/**
	 * @return data of the matrix
	 */
	protected String[][] getData() {
		return data;
	}
	
	/**
	 * @return String representation of the object.
	 */
	@Override
	public String toString() {
		String res = "";
		for (int i = 0; i < nbRow; i++) {
			for (int j = 0; j < nbCol; j++) {
				res += data[i][j];
			}
			res += "\n";
		}
		return res;
	}
	
}
